scripts/cleanup.js

```diff
@@ -53,124 +67,116 @@ function getImages(region, repo, callback) {
-  });
+    // We want to leave the repository with MAX_IMAGES number of images in it.
+    let imagesToDelete;
+    let excessImages = images.length - MAX_IMAGES;
+    if (excessImages <= 0) return callback(null, []);
 
-  console.log('cruftDigests: ', cruftDigests.join(','));
-  console.log('deployDigests: ', deployDigests.join(','));
+    // First take old images from generic commits
+    imagesToDelete = []
+      .concat(generic.slice(0, excessImages));
 
-  let digests = [];
-  digests = digests.concat(cruftDigests.slice(0, (digests.length - (MAX_IMAGES - 1))));
+    // Determine whether we still need to delete more images
+    excessImages = excessImages - imagesToDelete.length;
+    if (excessImages <= 0) return callback(null, imagesToDelete);
 
-  if (deployDigests.length > MAX_PRIORITY_IMAGES)
-    digests = digests.concat(deployDigests.slice(0, (deployDigests.length - MAX_PRIORITY_IMAGES)));
+    // Make sure that we always leave at least 50 priority images
+    excessImages = Math.max(
+      Math.min(priority.length - MIN_PRIORITY_IMAGES, excessImages),
+      0
+    );
 
-  console.log('digests ', digests.join(','));
-  return callback(null, digests);
+    // Select the oldest priority images for removal
+    imagesToDelete = imagesToDelete
+      .concat(priority.slice(0, excessImages));
+
+    return callback(null, imagesToDelete);
+  });
 }
 
 module.exports.deleteImages = deleteImages;
-function deleteImages(region, repo, imageIds, callback) {
+function deleteImages(region, repositoryName, images, callback) {
+  const ecr = new AWS.ECR({ region: region });
 
-  let ecr = new AWS.ECR({ region: region });
-  ecr.batchDeleteImage({
-    imageIds: imageIds,
-    repositoryName: repo
-  }, callback);
+  // Must delete images in batches of 100 max
+  const remaining = JSON.parse(JSON.stringify(images));
+  const imageIds = remaining.splice(0, 100);
 
+  if (!imageIds.length) return callback();
+
+  ecr.batchDeleteImage({ imageIds, repositoryName }, (err) => {
+    if (err) return callback(err);
+    deleteImages(region, repositoryName, remaining, callback);
+  });
 }
 
 module.exports.commitType = commitType;
-function commitType(sha, digest, callback) {
-  console.log('sha ', sha);
-  const spawn = require('child_process').spawn;
-
-  let type = {}; type[digest] = '';
-  function shspawn(command, callback) {
-    let cmd = spawn('sh', ['-c', command]);
-    let stdout, stderr;
-    cmd.stderr.on('data', err => {
-      stderr = err.toString('utf-8').trim();
-    });
-    cmd.stdout.on('data', out => {
-      stdout = out.toString('utf-8').trim();
-    });
-    cmd.on('close', () => {
-      // An error is just the git command erroring, which means that the
-      // output of the current command is void for determining type, as
-      // opposed to a true error in the script. So, just return null,
-      // instead of erroring.
-      if (stderr && stderr.length) return callback(null, null);
-      else return callback(null, stdout);
```

Comment:
You've written here a `spawn` wrapper that is essentially the same as `exec`. See [node.js docs](https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback).

The great thing about `exec` over `spawn` is that you get to just write a shell command, complete with pipes and whatever else. You also get `stdout` and `stderr` returned to you for free -- no need to collect it yourself.

Reply:
woah. O.O